### PR TITLE
PMI: add and wrap pmi bcast/allgather functions

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Wrapping the PMI usage in `MPIDU_shm_alloc` and `MPIDU_bc_table_create` into using `MPIR_pmi_bcast` and `MPIR_pmi_allgather`.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Passes tests (included warning check)
* [ ] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [ ] Commits are self-contained and do not do two things at once
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/warnings

**Comment by hzhou** [comment]
test:mpich/warnings

**Comment by hzhou** [comment]
test:mpich/warnings

**Comment by hzhou** [comment]
test:mpich/warnings

**Comment by hzhou** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch3/tcp

**Comment by hzhou** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch3/tcp

**Comment by hzhou** [comment]
test:jenkins/ch4/ofi

**Comment by hzhou** [comment]
test:jenkins/ch4/ofi

**Comment by hzhou** [comment]
test:jenkins/ch4/ofi

**Comment by hzhou** [comment]
test:jenkins/ch4/ofi

**Comment by hzhou** [comment]
test:jenkins/ch4/ofi

**Comment by hzhou** [comment]
test:jenkins/ch4/ofi

**Comment by hzhou** [comment]
test:jenkins/ch4/ofi

**Comment by hzhou** [comment]
test:jenkins/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/warnings
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:mpich/warnings
test:jenkins/ch4/all

**Comment by hzhou** [comment]
test:mpich/warnings
test:jenkins/ch4/pmi

**Comment by hzhou** [comment]
test:jenkins/custom
netmod=ch4:ofi
env: MPIR_CVAR_CH4_ROOTS_ONLY_PMI=1

**Comment by hzhou** [comment]
test:jenkins/custom
netmod=ch4:ofi
env: MPIR_CVAR_CH4_ROOTS_ONLY_PMI=1

**Comment by hzhou** [comment]
test:jenkins/custom
netmod=ch4:ofi
config=debug
env: MPIR_CVAR_CH4_ROOTS_ONLY_PMI=1

**Comment by hzhou** [comment]
test:jenkins/custom
netmod=ch4:ofi
config=debug
env: MPIR_CVAR_CH4_ROOTS_ONLY_PMI=1

**Comment by hzhou** [comment]
test:jenkins/custom
netmod=ch4:ofi
config=debug
env: MPIR_CVAR_CH4_ROOTS_ONLY_PMI=1

**Comment by hzhou** [comment]
test:jenkins/custom
netmod=ch4:ofi
config=debug
env: MPIR_CVAR_CH4_ROOTS_ONLY_PMI=1

**Comment by hzhou** [comment]
test:jenkins/custom
netmod=ch4:ofi
config=debug
env: MPIR_CVAR_CH4_ROOTS_ONLY_PMI=1

**Comment by hzhou** [comment]
test:mpich/warnings/more

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most
test:jenkins/ch4/pmi

**Comment by hzhou** [comment]
test:mpich/warnings/more
test:jenkins/ch3/most
test:jenkins/ch4/most
test:jenkins/ch4/pmi

test:jenkins/custom
netmod=ch4:ofi
config=debug
env: MPIR_CVAR_CH4_ROOTS_ONLY_PMI=1

**Comment by hzhou** [comment]
test:mpich/warnings/more

test:jenkins/custom
netmod=ch4:ofi
config=debug
env: MPIR_CVAR_CH4_ROOTS_ONLY_PMI=1

File: src/mpid/common/shm/mpidu_shm_alloc.c [hunk 1]
```diff
@@ -195,32 +189,29 @@ int MPIDU_shm_seg_commit(MPIDU_shm_seg_t * memory, MPIDU_shm_barrier_t ** barrie
...
         ipc_lock = (OPA_emulation_ipl_t *) ((char *) memory->base_addr + ipc_lock_offset);
         ret = OPA_Interprocess_lock_init(ipc_lock, TRUE /*isLeader */);
         MPIR_ERR_CHKANDJUMP1(ret != 0, mpi_errno, MPI_ERR_OTHER, "**fail", "**fail %d", ret);
 #endif
-
         mpi_errno =
             MPIDU_shm_barrier_init((MPIDU_shm_barrier_t *) memory->base_addr, barrier, TRUE);
         MPIR_ERR_CHECK(mpi_errno);
     } else {
-        MPIR_CHKLMEM_MALLOC(key, char *, PMI2_MAX_KEYLEN, mpi_errno, "key", MPL_MEM_SHM);
```
Comment (gcongiu) [inline src/mpid/common/shm/mpidu_shm_alloc.c]: Need to decrement count in `MPIR_CHKLMEM_DECL`

File: src/mpid/common/shm/mpidu_shm_alloc.c [hunk 2]
```diff
@@ -406,130 +263,18 @@ int MPIDU_shm_seg_commit(MPIDU_shm_seg_t * memory, MPIDU_shm_barrier_t ** barrie
                 MPIDU_shm_barrier_init((MPIDU_shm_barrier_t *) memory->base_addr, barrier, FALSE);
             MPIR_ERR_CHECK(mpi_errno);
         }
-
```
Comment (gcongiu) [inline src/mpid/common/shm/mpidu_shm_alloc.c]: Unrelated line removal.

File: src/mpid/common/shm/mpidu_shm_alloc.c (hunk 1 above)
Comment (gcongiu) [inline src/mpid/common/shm/mpidu_shm_alloc.c]: Need to decrement count in `MPIR_CHKLMEM_DECL`

File: src/mpid/common/shm/mpidu_shm_alloc.c (hunk 2 above)
Comment (gcongiu) [inline src/mpid/common/shm/mpidu_shm_alloc.c]: Unrelated line removal.

File: src/util/mpir_pmi.c [hunk 3]
```diff
@@ -187,7 +187,7 @@ int MPIR_pmi_kvs_put(const char *key, const char *val)
 #elif defined(USE_PMIX_API)
     pmix_value_t value;
     value.type = PMIX_STRING;
-    value.data.string = val;
+    value.data.string = (char *) val;
```
Comment (raffenet) [inline src/util/mpir_pmi.c]: These changes to `MPIR_pmi_kvs_put` and `MPIR_pmi_barrier_local` seem to be unrelated to the bcast and allgather implementations.

File: src/util/mpir_pmi.c [hunk 4]
```diff
@@ -557,3 +557,38 @@ static int build_locality(void)
 
     return MPI_SUCCESS;
 }
+
+/* similar to functions in mpl/src/str/mpl_argstr.c, but much simpler */
+static int hex(unsigned char c)
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/util/mpir_pmi.c]: We could. I am thinking to delay it in a separate PR where we find more places that also use this function. If only a single place is using it, moving to MPL seems simply adding abstractions.

[the remark this answers is not in the document]
Comment (raffenet) [reply src/util/mpir_pmi.c]: That works for me. Doesn't need to hold up this PR.

File: src/util/mpir_pmi.c (hunk 3 above)
Comment (hzhou) [reply src/util/mpir_pmi.c]: The fix to MPIR_pmi_kvs_put is a bug fix (for `PMIx`). 

`MPIR_pmi_barrier_local` originally was only used in `MPIDU_shm_seg_alloc` (that is the reason I refactor it out), but later turns out we don't really need it. `PMI1/2` don't support barrier local anyway. `PMIx` will block and wait for the key anyway. There are also some `MPIR_pmi_` functions in the end may not be needed, e.g. seems we only need `MPIR_pmi_allgather` for the node roots. So I am thinking leaving the interface pruning to a later PR when code stabilizes.

I could split the PR is you insist.

Comment (raffenet) [reply src/util/mpir_pmi.c]: I am reviewing the changes in this commit. The commit message says `pmi: implement MPIR_pmi_bcast and MPIR_pmi_allgather`, yet it is changing this functions and I do not understand why. This either should be explained in the commit message, or they should be moved to another commit explaining why they are being changed. They do not need to move to another PR, but they do not appear to belong in this commit.

Comment (hzhou) [reply src/util/mpir_pmi.c]: I see. It is a mix-up during warnings fix. I'll correct it.

File: src/mpid/common/shm/mpidu_shm_alloc.c (hunk 1 above)
Comment (hzhou) [reply src/mpid/common/shm/mpidu_shm_alloc.c]: fixed.

File: src/mpid/common/shm/mpidu_shm_alloc.c (hunk 2 above)
Comment (hzhou) [reply src/mpid/common/shm/mpidu_shm_alloc.c]: fixed.

File: src/util/mpir_pmi.c (hunk 3 above)
Comment (hzhou) [reply src/util/mpir_pmi.c]: fixed.

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most
test:jenkins/ch4/pmi
test:mpich/warnings

**Comment by hzhou** [comment]
@raffenet Could you take another look?
